### ISSUE TITLE
jsonb fields not working

### ISSUE THREAD

**Issue by stlbucket** [graphile/graphile-build#85]
has something changed with jsonb fields?
i have just moved from:
"postgraphql": "^4.0.0-alpha1.3"
to:
"postgraphile": "^4.0.0-alpha2.20",
and now my json fields don't seem to be going thru to my postgres functions. no errors, just empty

i've isolated as follows:

create a function:
```
create function myschema.ping_json(
  _the_json jsonb default '[]'
) returns jsonb as $$
declare
begin
  RETURN _the_json;
end;
$$ language plpgsql strict security definer;
--||--
GRANT execute ON FUNCTION soro.ping_json(
  jsonb
) TO myschema_user;
```
calling this directly works fine:
```
select myschema.ping_json(
 '[{"amount":"44"},{"amount":null}]'
)
```
but calling through graphql:
```
mutation PingJson($theJson: JSON!) {
  pingJson(
    input: {
      _theJson: $theJson
    }
  ) {
    json
  }
}
```
with variables:
```
{
 "theJson": "[{\"amount\":\"44\"},{\"amount\":null}]"
}
```
gives error:
```
{
  "errors": [
    {
      "message": "malformed array literal: \"{{}}\"",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "pingJson"
      ],
      "stack": "error: malformed array literal: \"{{}}\"\n    at Connection.parseE (/Users/buckfactor/soro/bitbucket/soro-quoting-post-graph-ql/node_modules/pg/lib/connection.js:567:11)\n    at Connection.parseMessage (/Users/buckfactor/soro/bitbucket/soro-quoting-post-graph-ql/node_modules/pg/lib/connection.js:391:17)\n    at Socket.<anonymous> (/Users/buckfactor/soro/bitbucket/soro-quoting-post-graph-ql/node_modules/pg/lib/connection.js:129:22)\n    at emitOne (events.js:115:13)\n    at Socket.emit (events.js:210:7)\n    at addChunk (_stream_readable.js:252:12)\n    at readableAddChunk (_stream_readable.js:239:11)\n    at Socket.Readable.push (_stream_readable.js:197:10)\n    at TCP.onread (net.js:589:20)"
    }
  ],
  "data": {
    "pingJson": null
  }
}
```
![image](https://user-images.githubusercontent.com/2467892/30969584-74508910-a428-11e7-8a42-9b5d6113d41b.png)

mutation wrapper:
```
const clog = require('fbkt-clog')

module.exports = function CreateQuoteLineItemWrapPlugin (builder) {
  builder.hook(
    "GraphQLObjectType:fields:field",
    (field,
     {pgSql: sql},
     {scope: {isRootMutation, fieldName}, addArgDataGenerator}) => {
      if (!isRootMutation || fieldName !== "pingJson") {
        // If it's not the root mutation, or the mutation isn't the 'createLink'
        // mutation then we don't want to modify it - so return the input object
        // unmodified.
        return field;
      }

      // We're going to need link.id for our `performAnotherTask`; so we're going
      // to abuse addArgDataGenerator to make sure that this field is ALWAYS
      // requested, even if the user doesn't specify it. We're careful to alias
      // the result to a field that begins with `__` as that's forbidden by
      // GraphQL and thus cannot clash with a user's fields.
      // addArgDataGenerator(() => ({
      //   pgQuery: queryBuilder => {
      //     queryBuilder.select(
      //       // Select this value from the result of the INSERT:
      //       sql.query`${queryBuilder.getTableAlias()}.id`,
      //       // And give it this name in the result data:
      //       "__createdRecordId"
      //     );
      //   },
      // }));

      // It's possible that `resolve` isn't specified on a field, so in that case
      // we fall back to a default resolver.
      const defaultResolver = obj => obj[fieldName];

      // Extract the old resolver from `field`
      const {resolve: oldResolve = defaultResolver, ...rest} = field;

      return {
        // Copy over everything except 'resolve'
        ...rest,

        // Add our new resolver which wraps the old resolver
        async resolve(...resolveParams) {
          // Perform some validation (or any other action you want to do before
          // calling the old resolver)
          // const RESOLVE_ARGS_INDEX = 1;
          // const {input: {link: {title}}} = resolveParams[
          //   RESOLVE_ARGS_INDEX
          //   ];
          // if (title.length < 3) {
          //   throw new Error("Title is too short!");
          // }
clog('PING JSON PARAMS', resolveParams[3].variableValues)
          // Call the old resolver (you SHOULD NOT modify the arguments it
          // receives unless you also manipulate the AST it gets passed as the
          // 4th argument; which is quite a lot of effort) and store the result.
          const oldResolveResult = await oldResolve(...resolveParams);

          // Perform any tasks we want to do after the record is created.
          // await performAnotherTask(oldResolveResult.data.__createdRecordId);
          // await handler(oldResolveResult);

          // Finally return the result.
          return oldResolveResult;
        },
      };
    }
  );
};
```
gives me:
![image](https://user-images.githubusercontent.com/2467892/30969892-55807648-a429-11e7-954b-abd6526d132f.png)

and the error is thrown from the resolver

**Comment by stlbucket** [comment]
after rolling back the server version to:
"postgraphql": "^4.0.0-alpha1.3"

![image](https://user-images.githubusercontent.com/2467892/30971144-c17a181a-a42c-11e7-8ef7-97de2b17255b.png)

**Comment by benjie** [comment]
Yep this definitely looks like a bug. I wonder if we're treating the JSON field as an array and thus casting it to an array within PG which PG is not appreciating - I know our handling of arrays has changed. Odd that this didn't show up in our JSON acceptance tests.

**Comment by benjie** [comment]
I've reproduced this 👍

**Comment by benjie** [comment]
I've now reproduced this in tests (which took a bit longer!). There seems to be two issues at play here: 1. adding the default argument to the function seems to be causing some issues (according to the SQL dump we're literally not passing the value to the function in this case?!!); 2. when we split mutations to a two-phase execution (so that side-effects had time to take place) we added logic to pass the data back up to the DB like v3 does. Unfortunately this seems to be broken in the case of a JSON array in some cases.

Work so far is on `fix-85` branch; this is definitely an unpleasant bug.

**Comment by benjie** [comment]
Sorry it took so long for me to get back to this. I've determined the issue - we were receiving parsed JSON from the `pg` module, but when posting that back through to the `pg` module verbatim it was not handling it correctly. Instead I have performed JSON.stringify in this interrim period to solve it. I'm expecting there may be other issues from this where pg-types parses other values - we really need to tell `pg` not to perform any parsing for this particular query.
